### PR TITLE
api/types/container: refactor to use strings.Cut, DRY, move tests and fix validation

### PR DESCRIPTION
**Pull request by thaJeztah**
### hostconfig: move unit tests to api/types/container

These types were moved to api/types/container in https://github.com/moby/moby/commit/7ac4232e70fe7cf7318333cd0890db7f95663079 (https://github.com/moby/moby/pull/18762), but the unit-tests for them were not moved. This patch moves the unit-tests back together with the types.

### api/types/container: fix validation for UTSMode, UsernsMode, PidMode

The IPCMode type was added in 497fc8876ede9924f61c0eee4dfadd71e5d9f537 (https://github.com/moby/moby/pull/9074), and from that patch, the intent was to allow `host` (without `:`), `""` (empty, default) or `container:<container ID>`, but the `Valid()` function seems to be too relaxed and accepting both `:`, as well as `host:<anything>`. No unit-tests were added in that patch, and integration-tests only tested for valid values.

Later on, `PidMode`, and `UTSMode` were added in 23feaaa240853c0e7f9817f8c2d272dd1c93ac3f ( https://github.com/moby/moby/pull/9339 / https://github.com/moby/moby/pull/10080) and f2e5207fc989288ad136d48222df8e7754eb0e9b (https://github.com/moby/moby/pull/12667), both of which were implemented as a straight copy of the `IPCMode` implementation, copying the same bug.

Finally, commit d4aec5f0a680b6b01bb720830451a93c6ec398e6 (https://github.com/moby/moby/pull/14134) implemented unit-tests for these types, but testing for the wrong behavior of the implementation.

This patch updates the validation to correctly invalidate `host[:<anything>]` and empty (`:`) types.

### CODE REVIEW

**Comment by thaJeztah** [comment]
@rumpl @corhere PTAL

**Review by corhere — APPROVED** [review]
Per [Hyrum's Law](https://www.hyrumslaw.com/) there is inevitably going to be someone (likely with a buggy script) who is going to be upset that we "broke" them. We should call this fix out in the changelog since it is technically a breaking API change.

File: api/types/container/hostconfig.go [hunk 1]
```diff
@@ -463,3 +432,13 @@ type HostConfig struct {
...
+
+// containerID splits "container:<ID|name>" values. It returns the container
+// ID or name, and whether an ID/name was found. It returns an empty string and
+// a "false" if the value does not have a "container:" prefix. Further validation
+// of the returned, including checking if the value is empty, should be handled
+// by the caller.
+func containerID(val string) (idOrName string, ok bool) {
+	k, idOrName, ok := strings.Cut(val, ":")
+	return idOrName, ok && k == "container"
+}
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply api/types/container/hostconfig.go:444]: Didn't want to go all overboard on DRY 😂 I think I'll keep that for a future exercise;

File: api/types/container/hostconfig.go [hunk 2]
```diff
@@ -183,36 +180,27 @@ func (n UsernsMode) IsPrivate() bool {
...
 // Valid indicates whether the userns is valid.
 func (n UsernsMode) Valid() bool {
-	parts := strings.Split(string(n), ":")
-	switch mode := parts[0]; mode {
-	case "", "host":
-	default:
-		return false
-	}
-	return true
+	return string(n) == "" || n.IsHost()
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply api/types/container/hostconfig.go:183]: LOL, I was actually looking at those after I rebased and pushed ("was this actually needed?"); I'll have a look.

**Comment by thaJeztah** [comment]
> Per [Hyrum's Law](https://www.hyrumslaw.com/) there is inevitably going to be someone (likely with a buggy script) who is going to be upset that we "broke" them. We should call this fix out in the changelog since it is technically a breaking API change.

@corhere I added `impact/changelog`, and removed the redundant `string()` casts; PTAL 🤗

File: api/types/container/hostconfig.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (corhere) [reply api/types/container/hostconfig.go:444]: Fair. I only really looked into this to sate my curiosity.

**Review by corhere — APPROVED** [review]

**Pull request merged**